File: src/main/java/net/flyingfishflash/loremlist/api/LrmListApiServiceDefault.java
package net.flyingfishflash.loremlist.api;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;
import kotlin.Pair;
import kotlin.Triple;
import net.flyingfishflash.loremlist.api.data.request.LrmItemCreateRequest;
import net.flyingfishflash.loremlist.api.data.request.LrmListCreateRequest;
import net.flyingfishflash.loremlist.api.data.response.ApiServiceResponse;
import net.flyingfishflash.loremlist.api.data.response.AssociationDeletedResponse;
import net.flyingfishflash.loremlist.api.data.response.AssociationsDeletedResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmItemResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmListDeletedResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmListItemAddedResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmListItemMovedResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmListItemResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmListResponse;
import net.flyingfishflash.loremlist.core.response.structure.ApiMessageNumeric;
import net.flyingfishflash.loremlist.domain.ServiceResponse;
import net.flyingfishflash.loremlist.domain.lrmitem.data.LrmItemCreate;
import net.flyingfishflash.loremlist.domain.lrmlist.LrmList;
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListService;
import net.flyingfishflash.loremlist.domain.lrmlist.data.LrmListCreate;
import net.flyingfishflash.loremlist.domain.lrmlist.data.LrmListDeleted;
import net.flyingfishflash.loremlist.domain.lrmlistitem.LrmListItem;
import net.flyingfishflash.loremlist.domain.lrmlistitem.LrmListItemService;
import net.flyingfishflash.loremlist.domain.lrmlistitem.data.LrmListItemAdded;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
public class LrmListApiServiceDefault implements LrmListApiService {

  private final LrmListService lrmListService;
  private final LrmListItemService lrmListItemService;

  public LrmListApiServiceDefault(
      LrmListService lrmListService, LrmListItemService lrmListItemService) {
    this.lrmListService = lrmListService;
    this.lrmListItemService = lrmListItemService;
  }

  @Override
  public ApiServiceResponse<ApiMessageNumeric> countByOwner(String owner) {
    ServiceResponse<Long> serviceResponse = lrmListService.countByOwner(owner);
    return new ApiServiceResponse<>(
        new ApiMessageNumeric(serviceResponse.getContent()), serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<LrmListResponse> create(
      LrmListCreateRequest lrmListCreateRequest, String owner) {
    // TODO: evaluate where LrmItemRequest or LrmItem should be provided to domain service
    LrmListCreate lrmListCreate =
        new LrmListCreate(
            lrmListCreateRequest.getName(),
            lrmListCreateRequest.getDescription(),
            lrmListCreateRequest.getPublic());
    ServiceResponse<LrmList> serviceResponse = lrmListService.create(lrmListCreate, owner);
    return new ApiServiceResponse<>(
        LrmListResponse.fromLrmList(serviceResponse.getContent()), serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<LrmListDeletedResponse> deleteByOwner(String owner) {
    ServiceResponse<LrmListDeleted> serviceResponse = lrmListService.deleteByOwner(owner);
    LrmListDeleted content = serviceResponse.getContent();
    return new ApiServiceResponse<>(
        new LrmListDeletedResponse(content.listNames(), content.associatedItemNames()),
        serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<LrmListDeletedResponse> deleteByOwnerAndId(
      UUID id, String owner, boolean removeItemAssociations) {
    ServiceResponse<LrmListDeleted> serviceResponse =
        lrmListService.deleteByOwnerAndId(id, owner, removeItemAssociations);
    LrmListDeleted content = serviceResponse.getContent();
    return new ApiServiceResponse<>(
        new LrmListDeletedResponse(content.listNames(), content.associatedItemNames()),
        serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<List<LrmItemResponse>> findEligibleItemsByOwner(
      UUID listId, String owner) {
    throw new UnsupportedOperationException("Not yet implemented");
  }

  @Override
  public ApiServiceResponse<List<LrmListResponse>> findByOwner(String owner) {
    ServiceResponse<List<LrmList>> serviceResponse = lrmListService.findByOwner(owner);
    List<LrmListResponse> content =
        serviceResponse.getContent().stream().map(LrmListResponse::fromLrmList).toList();
    return new ApiServiceResponse<>(content, serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<List<LrmListResponse>> findByOwnerExcludeItems(String owner) {
    ServiceResponse<List<LrmList>> serviceResponse = lrmListService.findByOwner(owner);
    List<LrmListResponse> content =
        serviceResponse.getContent().stream()
            .map(list -> LrmListResponse.fromLrmList(list.withItems(Set.of())))
            .toList();
    return new ApiServiceResponse<>(content, serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<List<LrmListResponse>> findByOwnerAndHavingNoItemAssociations(
      String owner) {
    ServiceResponse<List<LrmList>> serviceResponse =
        lrmListService.findByOwnerAndHavingNoItemAssociations(owner);
    List<LrmListResponse> content =
        serviceResponse.getContent().stream().map(LrmListResponse::fromLrmList).toList();
    return new ApiServiceResponse<>(content, serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<LrmListResponse> findByOwnerAndId(UUID id, String owner) {
    ServiceResponse<LrmList> serviceResponse = lrmListService.findByOwnerAndId(id, owner);
    return new ApiServiceResponse<>(
        LrmListResponse.fromLrmList(serviceResponse.getContent()), serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<LrmListResponse> findByOwnerAndIdExcludeItems(UUID id, String owner) {
    ServiceResponse<LrmList> serviceResponse = lrmListService.findByOwnerAndId(id, owner);
    return new ApiServiceResponse<>(
        LrmListResponse.fromLrmList(serviceResponse.getContent().withItems(Set.of())),
        serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<List<LrmListResponse>> findByPublic() {
    ServiceResponse<List<LrmList>> serviceResponse = lrmListService.findByPublic();
    List<LrmListResponse> content =
        serviceResponse.getContent().stream().map(LrmListResponse::fromLrmList).toList();
    return new ApiServiceResponse<>(content, serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<List<LrmListResponse>> findByPublicExcludeItems() {
    ServiceResponse<List<LrmList>> serviceResponse = lrmListService.findByPublic();
    List<LrmListResponse> content =
        serviceResponse.getContent().stream()
            .map(list -> LrmListResponse.fromLrmList(list.withItems(Set.of())))
            .toList();
    return new ApiServiceResponse<>(content, serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<LrmListResponse> patchByOwnerAndId(
      UUID id, String owner, Map<String, Object> patchRequest) {
    LrmList lrmList = lrmListService.findByOwnerAndId(id, owner).getContent();
    Set<String> patchedFields = new LinkedHashSet<>();
    for (Map.Entry<String, Object> entry : patchRequest.entrySet()) {
      String fieldToPatch = entry.getKey();
      Object value = entry.getValue();
      switch (fieldToPatch) {
        case "name" -> {
          if (!Objects.equals(value, lrmList.name())) {
            lrmListService.patchName(lrmList.withName((String) value));
            patchedFields.add(fieldToPatch);
          }
        }
        case "description" -> {
          if (!Objects.equals(value, lrmList.description())) {
            lrmListService.patchDescription(lrmList.withDescription((String) value));
            patchedFields.add(fieldToPatch);
          }
        }
        case "public" -> {
          if (!Objects.equals(value, lrmList.isPublic())) {
            lrmListService.patchIsPublic(lrmList.withIsPublic((Boolean) value));
            patchedFields.add(fieldToPatch);
          }
        }
        default ->
            throw new IllegalArgumentException(
                "Patch operation is not supported on field: " + fieldToPatch);
      }
    }
    LrmList patchedLrmList = lrmListService.findByOwnerAndId(id, owner).getContent();
    String message;
    if (!patchedLrmList.equals(lrmList)) {
      message =
          "List '"
              + patchedLrmList.name()
              + "' updated. Fields changed: "
              + String.join(", ", patchedFields)
              + ".";
    } else {
      message = "List '" + patchedLrmList.name() + "' not updated.";
    }
    LrmListResponse lrmItemResponse = LrmListResponse.fromLrmList(patchedLrmList);
    return new ApiServiceResponse<>(lrmItemResponse, message);
  }

  // list item context

  @Override
  public ApiServiceResponse<LrmListItemAddedResponse> addListItem(
      UUID listId, Set<UUID> itemIdCollection, String owner) {
    ServiceResponse<LrmListItemAdded> serviceResponse =
        lrmListItemService.add(listId, new java.util.ArrayList<>(itemIdCollection), owner);
    LrmListItemAdded content = serviceResponse.getContent();
    LrmListItemAddedResponse lrmListItemAddedResponse =
        new LrmListItemAddedResponse(content.listName(), content.items());
    return new ApiServiceResponse<>(lrmListItemAddedResponse, serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<ApiMessageNumeric> countListItems(UUID listId, String listOwner) {
    ServiceResponse<Long> serviceResponse =
        lrmListItemService.countByOwnerAndListId(listId, listOwner);
    return new ApiServiceResponse<>(
        new ApiMessageNumeric(serviceResponse.getContent()), serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<LrmListItemResponse> createListItem(
      UUID listId, LrmItemCreateRequest itemCreateRequest, String creator) {
    LrmItemCreate lrmItemCreate =
        new LrmItemCreate(
            itemCreateRequest.getName(),
            itemCreateRequest.getDescription(),
            itemCreateRequest.getQuantity(),
            itemCreateRequest.isSuppressed());
    ServiceResponse<LrmListItem> serviceResponse =
        lrmListItemService.create(listId, lrmItemCreate, creator);
    return new ApiServiceResponse<>(
        LrmListItemResponse.fromLrmListItem(serviceResponse.getContent()),
        serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<LrmListItemResponse> findListItem(
      UUID listId, UUID itemId, String listOwner) {
    ServiceResponse<LrmListItem> serviceResponse =
        lrmListItemService.findByOwnerAndItemIdAndListId(itemId, listId, listOwner);
    return new ApiServiceResponse<>(
        LrmListItemResponse.fromLrmListItem(serviceResponse.getContent()),
        serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<LrmListItemMovedResponse> moveListItem(
      UUID listId, UUID itemId, UUID destinationListId, String owner) {
    ServiceResponse<Triple<String, String, String>> serviceResponse =
        lrmListItemService.move(itemId, listId, destinationListId, owner);
    Triple<String, String, String> content = serviceResponse.getContent();
    LrmListItemMovedResponse lrmListItemMovedResponse =
        new LrmListItemMovedResponse(content.getFirst(), content.getSecond(), content.getThird());
    return new ApiServiceResponse<>(lrmListItemMovedResponse, serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<AssociationDeletedResponse> removeListItem(
      UUID listId, UUID itemId, String componentsOwner) {
    ServiceResponse<Pair<String, String>> serviceResponse =
        lrmListItemService.removeByOwnerAndListIdAndItemId(listId, itemId, componentsOwner);
    Pair<String, String> content = serviceResponse.getContent();
    AssociationDeletedResponse listItemRemovedResponse =
        new AssociationDeletedResponse(content.getFirst(), content.getSecond());
    return new ApiServiceResponse<>(listItemRemovedResponse, serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<AssociationsDeletedResponse> removeAllListItems(
      UUID listId, String listOwner) {
    ServiceResponse<Pair<String, Integer>> serviceResponse =
        lrmListItemService.removeByOwnerAndListId(listId, listOwner);
    Pair<String, Integer> content = serviceResponse.getContent();
    AssociationsDeletedResponse listItemsRemovedResponse =
        new AssociationsDeletedResponse(content.getFirst(), content.getSecond());
    return new ApiServiceResponse<>(listItemsRemovedResponse, serviceResponse.getMessage());
  }

  @Override
  public ApiServiceResponse<LrmListItemResponse> patchListItem(
      UUID listId, UUID itemId, String listOwner, Map<String, Object> patchRequest) {
    LrmListItem lrmListItem =
        lrmListItemService.findByOwnerAndItemIdAndListId(itemId, listId, listOwner).getContent();
    Set<String> patchedFields = new LinkedHashSet<>();
    for (Map.Entry<String, Object> entry : patchRequest.entrySet()) {
      String fieldToPatch = entry.getKey();
      Object value = entry.getValue();
      switch (fieldToPatch) {
        case "name" -> {
          if (!Objects.equals(value, lrmListItem.name())) {
            lrmListItemService.patchName(lrmListItem.withName((String) value));
            patchedFields.add(fieldToPatch);
          }
        }
        case "description" -> {
          if (!Objects.equals(value, lrmListItem.description())) {
            lrmListItemService.patchDescription(lrmListItem.withDescription((String) value));
            patchedFields.add(fieldToPatch);
          }
        }
        case "quantity" -> {
          if (!Objects.equals(value, lrmListItem.quantity())) {
            lrmListItemService.patchQuantity(lrmListItem.withQuantity((Integer) value));
            patchedFields.add(fieldToPatch);
          }
        }
        case "isSuppressed" -> {
          if (!Objects.equals(value, lrmListItem.isSuppressed())) {
            lrmListItemService.patchIsSuppressed(lrmListItem.withIsSuppressed((Boolean) value));
            patchedFields.add(fieldToPatch);
          }
        }
        default ->
            throw new IllegalArgumentException(
                "Patch operation is not supported on field: " + fieldToPatch);
      }
    }

    LrmListItem patchedLrmListItem =
        lrmListItemService.findByOwnerAndItemIdAndListId(itemId, listId, listOwner).getContent();
    String message;
    if (!patchedLrmListItem.equals(lrmListItem)) {
      message =
          "Item '"
              + patchedLrmListItem.name()
              + "' updated. Fields changed: "
              + String.join(", ", patchedFields)
              + ".";
    } else {
      message = "Item '" + patchedLrmListItem.name() + "' not updated.";
    }
    LrmListItemResponse lrmItemResponse = LrmListItemResponse.fromLrmListItem(patchedLrmListItem);
    return new ApiServiceResponse<>(lrmItemResponse, message);
  }
}

File: src/main/kotlin/net/flyingfishflash/loremlist/api/LrmListApiServiceDefault.kt
package net.flyingfishflash.loremlist.api

import net.flyingfishflash.loremlist.api.data.request.LrmItemCreateRequest
import net.flyingfishflash.loremlist.api.data.request.LrmListCreateRequest
import net.flyingfishflash.loremlist.api.data.response.ApiServiceResponse
import net.flyingfishflash.loremlist.api.data.response.AssociationDeletedResponse
import net.flyingfishflash.loremlist.api.data.response.AssociationsDeletedResponse
import net.flyingfishflash.loremlist.api.data.response.LrmItemResponse
import net.flyingfishflash.loremlist.api.data.response.LrmListDeletedResponse
import net.flyingfishflash.loremlist.api.data.response.LrmListItemAddedResponse
import net.flyingfishflash.loremlist.api.data.response.LrmListItemMovedResponse
import net.flyingfishflash.loremlist.api.data.response.LrmListItemResponse
import net.flyingfishflash.loremlist.api.data.response.LrmListResponse
import net.flyingfishflash.loremlist.core.response.structure.ApiMessageNumeric
import net.flyingfishflash.loremlist.domain.lrmitem.data.LrmItemCreate
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListService
import net.flyingfishflash.loremlist.domain.lrmlist.data.LrmListCreate
import net.flyingfishflash.loremlist.domain.lrmlistitem.LrmListItemService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
class LrmListApiServiceDefault(private val lrmListService: LrmListService, private val lrmListItemService: LrmListItemService) : LrmListApiService {

  override fun countByOwner(owner: String): ApiServiceResponse<ApiMessageNumeric> {
    val serviceResponse = lrmListService.countByOwner(owner)
    return ApiServiceResponse(content = ApiMessageNumeric(serviceResponse.content), message = serviceResponse.message)
  }

  override fun create(lrmListCreateRequest: LrmListCreateRequest, owner: String): ApiServiceResponse<LrmListResponse> {
    // TODO: evaluate where LrmItemRequest or LrmItem should be provided to domain service
    val lrmListCreate = LrmListCreate(name = lrmListCreateRequest.name, description = lrmListCreateRequest.description, public = lrmListCreateRequest.public)
    val serviceResponse = lrmListService.create(lrmListCreate, owner)
    return ApiServiceResponse(content = LrmListResponse.fromLrmList(serviceResponse.content), message = serviceResponse.message)
  }

  override fun deleteByOwner(owner: String): ApiServiceResponse<LrmListDeletedResponse> {
    val serviceResponse = lrmListService.deleteByOwner(owner)
    return ApiServiceResponse(
      content = LrmListDeletedResponse(
        listNames = serviceResponse.content.listNames,
        associatedItemNames = serviceResponse.content.associatedItemNames,
      ),
      message = serviceResponse.message,
    )
  }

  override fun deleteByOwnerAndId(id: UUID, owner: String, removeItemAssociations: Boolean): ApiServiceResponse<LrmListDeletedResponse> {
    val serviceResponse = lrmListService.deleteByOwnerAndId(id = id, owner = owner, removeItemAssociations = removeItemAssociations)
    return ApiServiceResponse(
      content = LrmListDeletedResponse(
        listNames = serviceResponse.content.listNames,
        associatedItemNames = serviceResponse.content.associatedItemNames,
      ),
      message = serviceResponse.message,
    )
  }

  override fun findEligibleItemsByOwner(listId: UUID, owner: String): ApiServiceResponse<List<LrmItemResponse>> {
    TODO("Not yet implemented")
  }

  override fun findByOwner(owner: String): ApiServiceResponse<List<LrmListResponse>> {
    val serviceResponse = lrmListService.findByOwner(owner)
    return ApiServiceResponse(content = serviceResponse.content.map { LrmListResponse.fromLrmList(it) }, message = serviceResponse.message)
  }

  override fun findByOwnerExcludeItems(owner: String): ApiServiceResponse<List<LrmListResponse>> {
    val serviceResponse = lrmListService.findByOwner(owner)
    return ApiServiceResponse(
      content = serviceResponse.content.map { LrmListResponse.fromLrmList(it.copy(items = emptySet())) },
      message = serviceResponse.message,
    )
  }

  override fun findByOwnerAndHavingNoItemAssociations(owner: String): ApiServiceResponse<List<LrmListResponse>> {
    val serviceResponse = lrmListService.findByOwnerAndHavingNoItemAssociations(owner)
    return ApiServiceResponse(content = serviceResponse.content.map { LrmListResponse.fromLrmList(it) }, message = serviceResponse.message)
  }

  override fun findByOwnerAndId(id: UUID, owner: String): ApiServiceResponse<LrmListResponse> {
    val serviceResponse = lrmListService.findByOwnerAndId(id = id, owner = owner)
    return ApiServiceResponse(content = LrmListResponse.fromLrmList(serviceResponse.content), message = serviceResponse.message)
  }

  override fun findByOwnerAndIdExcludeItems(id: UUID, owner: String): ApiServiceResponse<LrmListResponse> {
    val serviceResponse = lrmListService.findByOwnerAndId(id = id, owner = owner)
    return ApiServiceResponse(content = LrmListResponse.fromLrmList(serviceResponse.content.copy(items = emptySet())), message = serviceResponse.message)
  }

  override fun findByPublic(): ApiServiceResponse<List<LrmListResponse>> {
    val serviceResponse = lrmListService.findByPublic()
    return ApiServiceResponse(content = serviceResponse.content.map { LrmListResponse.fromLrmList(it) }, message = serviceResponse.message)
  }

  override fun findByPublicExcludeItems(): ApiServiceResponse<List<LrmListResponse>> {
    val serviceResponse = lrmListService.findByPublic()
    return ApiServiceResponse(
      content = serviceResponse.content.map { LrmListResponse.fromLrmList(it.copy(items = emptySet())) },
      message = serviceResponse.message,
    )
  }

  override fun patchByOwnerAndId(id: UUID, owner: String, patchRequest: Map<String, Any>): ApiServiceResponse<LrmListResponse> {
    val lrmList = lrmListService.findByOwnerAndId(id = id, owner = owner).content
    val patchedFields = mutableSetOf<String>()
    patchRequest.forEach { (fieldToPatch, value) ->
      when (fieldToPatch) {
        "name" -> {
          if (value != lrmList.name) {
            lrmListService.patchName(lrmList.copy(name = value as String))
            patchedFields.add(fieldToPatch)
          }
        }
        "description" -> {
          if (value != lrmList.description) {
            lrmListService.patchDescription(lrmList.copy(description = value as String))
            patchedFields.add(fieldToPatch)
          }
        }
        "public" -> {
          if (value != lrmList.public) {
            lrmListService.patchIsPublic(lrmList.copy(public = value as Boolean))
            patchedFields.add(fieldToPatch)
          }
        }
        else -> {
          throw IllegalArgumentException("Patch operation is not supported on field: $fieldToPatch")
        }
      }
    }
    val patchedLrmList = lrmListService.findByOwnerAndId(id = id, owner = owner).content
    val message = if (patchedLrmList != lrmList) {
      "List '${patchedLrmList.name}' updated. Fields changed: ${patchedFields.joinToString() + "."}"
    } else {
      "List '${patchedLrmList.name}' not updated."
    }
    val lrmItemResponse = LrmListResponse.fromLrmList(patchedLrmList)
    return ApiServiceResponse(
      content = lrmItemResponse,
      message = message,
    )
  }

// list item context

  override fun addListItem(listId: UUID, itemIdCollection: Set<UUID>, owner: String): ApiServiceResponse<LrmListItemAddedResponse> {
    val serviceResponse = lrmListItemService.add(id = listId, idCollection = itemIdCollection.toList(), componentsOwner = owner)
    val lrmListItemAddedResponse = LrmListItemAddedResponse(
      componentName = serviceResponse.content.listName,
      associatedComponents = serviceResponse.content.items,
    )
    return ApiServiceResponse(content = lrmListItemAddedResponse, message = serviceResponse.message)
  }

  override fun countListItems(listId: UUID, listOwner: String): ApiServiceResponse<ApiMessageNumeric> {
    val serviceResponse = lrmListItemService.countByOwnerAndListId(listId = listId, owner = listOwner)
    return ApiServiceResponse(content = ApiMessageNumeric(serviceResponse.content), message = serviceResponse.message)
  }

  override fun createListItem(listId: UUID, itemCreateRequest: LrmItemCreateRequest, creator: String): ApiServiceResponse<LrmListItemResponse> {
    val lrmItemCreate =
      LrmItemCreate(
        name = itemCreateRequest.name,
        description = itemCreateRequest.description,
        quantity = itemCreateRequest.quantity,
        isSuppressed = itemCreateRequest.isSuppressed,
      )

    val serviceResponse = lrmListItemService.create(listId = listId, lrmItemCreate, creator = creator)
    return ApiServiceResponse(content = LrmListItemResponse.fromLrmListItem(serviceResponse.content), message = serviceResponse.message)
  }

  override fun findListItem(listId: UUID, itemId: UUID, listOwner: String): ApiServiceResponse<LrmListItemResponse> {
    val serviceResponse = lrmListItemService.findByOwnerAndItemIdAndListId(listId = listId, itemId = itemId, owner = listOwner)
    return ApiServiceResponse(content = LrmListItemResponse.fromLrmListItem(serviceResponse.content), message = serviceResponse.message)
  }

  override fun moveListItem(listId: UUID, itemId: UUID, destinationListId: UUID, owner: String): ApiServiceResponse<LrmListItemMovedResponse> {
    val serviceResponse = lrmListItemService.move(
      currentListId = listId,
      itemId = itemId,
      destinationListId = destinationListId,
      owner = owner,
    )

    val lrmListItemMovedResponse = LrmListItemMovedResponse(
      itemName = serviceResponse.content.first,
      currentListName = serviceResponse.content.second,
      newListName = serviceResponse.content.third,
    )

    return ApiServiceResponse(content = lrmListItemMovedResponse, message = serviceResponse.message)
  }

  override fun removeListItem(listId: UUID, itemId: UUID, componentsOwner: String): ApiServiceResponse<AssociationDeletedResponse> {
    val serviceResponse = lrmListItemService.removeByOwnerAndListIdAndItemId(itemId = itemId, listId = listId, owner = componentsOwner)
    val listItemRemovedResponse = AssociationDeletedResponse(itemName = serviceResponse.content.first, listName = serviceResponse.content.second)
    return ApiServiceResponse(content = listItemRemovedResponse, message = serviceResponse.message)
  }

  override fun removeAllListItems(listId: UUID, listOwner: String): ApiServiceResponse<AssociationsDeletedResponse> {
    val serviceResponse = lrmListItemService.removeByOwnerAndListId(listId = listId, owner = listOwner)
    val listItemsRemovedResponse = AssociationsDeletedResponse(
      itemName = serviceResponse.content.first,
      deletedAssociationsCount = serviceResponse.content.second,
    )
    return ApiServiceResponse(content = listItemsRemovedResponse, message = serviceResponse.message)
  }

  override fun patchListItem(listId: UUID, itemId: UUID, listOwner: String, patchRequest: Map<String, Any>): ApiServiceResponse<LrmListItemResponse> {
    val lrmListItem = lrmListItemService.findByOwnerAndItemIdAndListId(itemId = itemId, listId = listId, owner = listOwner).content
    val patchedFields = mutableSetOf<String>()
    patchRequest.forEach { (fieldToPatch, value) ->
      when (fieldToPatch) {
        "name" -> {
          if (value != lrmListItem.name) {
            lrmListItemService.patchName(lrmListItem.copy(name = value as String))
            patchedFields.add(fieldToPatch)
          }
        }
        "description" -> {
          if (value != lrmListItem.description) {
            lrmListItemService.patchDescription(lrmListItem.copy(description = value as String))
            patchedFields.add(fieldToPatch)
          }
        }
        "quantity" -> {
          if (value != lrmListItem.quantity) {
            lrmListItemService.patchQuantity(lrmListItem.copy(quantity = value as Int))
            patchedFields.add(fieldToPatch)
          }
        }
        "isSuppressed" -> {
          if (value != lrmListItem.isSuppressed) {
            lrmListItemService.patchIsSuppressed(lrmListItem.copy(isSuppressed = value as Boolean))
            patchedFields.add(fieldToPatch)
          }
        }
        else -> {
          throw IllegalArgumentException("Patch operation is not supported on field: $fieldToPatch")
        }
      }
    }

    val patchedLrmListItem = lrmListItemService.findByOwnerAndItemIdAndListId(itemId = itemId, listId = listId, owner = listOwner).content
    val message = if (patchedLrmListItem != lrmListItem) {
      "Item '${patchedLrmListItem.name}' updated. Fields changed: ${patchedFields.joinToString() + "."}"
    } else {
      "Item '${patchedLrmListItem.name}' not updated."
    }
    val lrmItemResponse = LrmListItemResponse.fromLrmListItem(patchedLrmListItem)
    return ApiServiceResponse(content = lrmItemResponse, message = message)
  }
}
